File: scala/113.scala
object Solution {
    def pathSum(root: TreeNode, targetSum: Int): List[List[Int]] = {
        def dfs(root: TreeNode, path: List[Int], sum: Int): List[List[Int]] = {
          if (root == null) Nil
          else if (root.left == null && root.right == null) {
            if (sum+root.value == targetSum) List(path :+ root.value)
            else Nil
          } else {
            dfs(root.left, path :+ root.value, sum+root.value) ::: dfs(root.right, path :+ root.value, sum+root.value)
          }
        }
        dfs(root, Nil, 0)
    }
}

File: leetcodeJava/113.java
package leetcodeJava;

class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        var res = new LinkedList<List<Integer>>();
        if (root == null) return res;
        dfs(root, new LinkedList<TreeNode>(), 0, targetSum, res);
        return res;
    }
    
    private void dfs(TreeNode root, LinkedList<TreeNode> path, int sum, int target, List<List<Integer>> res) {
        path.add(root);
        sum = sum + root.val;
        if (sum == target && root.left == null && root.right == null) {
            res.add(path.stream().map(n -> Integer.valueOf(n.val)).collect(Collectors.toList()));
        }
        if (root.left != null)
            dfs(root.left, path, sum, target, res);
        if (root.right != null)
            dfs(root.right, path, sum, target, res);
        if (path.size() != 0)
            path.removeLast();
    }
}
